src/fairseq2/datasets/huggingface.py

```diff
@@ -0,0 +1,153 @@
+
+try:
+    from datasets import (  # type: ignore[attr-defined,import-untyped,import-not-found]
+        Dataset,
+        DatasetDict,
+    )
+except ImportError:
+    has_datasets = False
+else:
+    has_datasets = True
+
+
+Example = Dict[str, Any]
+
+
+def create_hf_reader(
+    dataset: Dataset,
+    gang: Gang,
+    converter: Callable[[Example], BatchT],
+    *,
+    batching: Optional[Union[StaticBatching, LengthBatching]] = None,
+    max_seq_len: Optional[int] = None,
+    drop_remainder: bool = False,
+    min_seq_len: int = 0,
+    seq_len_col: Optional[str] = None,
+    num_accumulate: int = 1,
+    num_prefetch: int = 1,
+    **extra: Any,
+) -> DataPipelineReader[BatchT]:
+    """
+    Convert HF Dataset into a fairseq2 PipelineReader.
+    The HF-specific processing logic is assumed to be
+    done before the conversion.
+
+    **NOTE**: This function should not be used for big datatset, but
+    as a convenient wrapper over some small HuggingFace datasets for
+    quick iteration.
+
+    **Batching**:
+
+    :param gang:
+        The gang over which to shard the dataset.
+    :param batching:
+        The batching strategy that is used to transform the pipeline
+        into iteration of batch of items. If None, The batching is
+        done externally using the datasets API.
+    :param drop_remainder:
+        Whether or not drop the remainder items that do not fit the
+        batching strategy (e.g. not having the batch size)
+    :param max_seq_len:
+        used when batching by length, to define the cap of the batch
+    :param min_seq_len:
+        used to filter the items with too short length
+    :param seq_len_col:
+        The column storing example's length
+    :param num_accumulate:
+        The number of batches to accumulate in each iteration.
+        Typically used with gradient accumulation during
+        training.
+    :param num_prefetch:
+        The number of batches to prefetch in background.
+    :param extras:
+        The extra parameters specific to the dataset
+        implementation.
+    """
+
+    if not has_datasets:
+        raise ModuleNotFoundError(
+            "`datasets` is required but not found. Please install it with `pip install datasets`."
+        )  # fmt: skip
+
+    # Make sure the dataset is a proper arrow dataset
+    if not isinstance(dataset, Dataset):
+        # One common mistake is pass a DatasetDict (e.g. with all splits) as inputs
+        if isinstance(dataset, DatasetDict):
+            raise TypeError(
+                "create_reader() expects input of datasets.Dataset type, "
+                "get datasets.DatasetDict. Make sure you specify `split` "
+                "when loading the dataset"
+            )
+        raise TypeError(f"Expect datasets.Dataset type, get {type(dataset)}")
+
+    if batching is None:
+        data = dataset.data.to_batches()  # list of RecordBatch
+    else:
+        data = dataset.to_list()
+    builder = read_sequence(data)
```

Comment:
Nit: we now have a new `read_iterator` function that might be more efficient than `read_sequence`.